src/ft60x.rs

```diff
@@ -7,151 +7,133 @@ use crate::ft60x_config::FT60xConfig;
-                while producer
-                    .with_next_buffer(|buf| {
-                        let mut async_group = AsyncGroup::new(&self.context);
-                        for (i, chunk) in buf.chunks_mut(blocksize).enumerate() {
-                            async_group
-                                .submit(Transfer::bulk(&device, 0x82, chunk, Duration::new(1, 0)))
-                                .unwrap();
-
-                            // The FT60x doesn't seem to like too many outstanding requests
-                            if i > 50 {
-                                assert_eq!(
-                                    async_group.wait_any().unwrap().actual().len(),
-                                    blocksize
-                                );
-                            }
-                        }
-
-                        while let Ok(mut transfer) = async_group.wait_any() {
-                            assert_eq!(transfer.actual().len(), blocksize);
-                        }
-                    })
-                    .is_ok()
-                {}
-
-                Ok(())
-            })
-            .unwrap();
-        });
-
-        Ok(consumer)
-    }
-
-    fn set_streaming_mode(&self) -> Result<()> {
-        self.with_device(|device| {
-            device.claim_interface(0)?;
-            device.claim_interface(1)?;
+    fn set_streaming_mode(&mut self) -> Result<()> {
+        if !self.streaming_mode {
+            self.device.claim_interface(0)?;
+            self.device.claim_interface(1)?;
 
             let ctrlreq = [
                 0x00, 0x00, 0x00, 0x00, 0x82, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x40, 0x00, 0x00,
                 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
             ];
 
-            device.write_bulk(0x01, &ctrlreq, Duration::new(1, 0))?;
+            self.device.write_bulk(0x01, &ctrlreq, Duration::new(1, 0))?;
+            self.streaming_mode = true;
+        }
+        Ok(())
+    }
 
-            Ok(())
-        })?;
+    /// it es recommended to read multiples of 32Kb
+    pub fn read_exact(&mut self, buf: &mut [u8]) -> Result<()> {
+        if buf.len() == 0 { return Ok(()); }
+
+        self.set_streaming_mode()?;
+
+        let blocksize: usize = 32 * 1024; // 32 Kb seems to be the sweet spot for the ft601
+        let mut_chunks = buf.chunks_mut(blocksize);
+        let mut_chunks_len = mut_chunks.len();
+        let mut collected = 0;
+        let mut last_len = 0;
+
+        let mut async_group = AsyncGroup::new(&self.context);
+        for (i, chunk) in mut_chunks.enumerate() {
+            if i == mut_chunks_len - 1 { last_len = chunk.len() }
+            async_group.submit(Transfer::bulk(&self.device, 0x82, chunk, Duration::new(1, 0)))?;
+
+            // The FT60x doesn't seem to like too many outstanding requests
+            if i > 50 {
+                let real_len = async_group.wait_any()?.actual().len();
```

Comment:
This is also invalid, the transfers could complete out of order (or atleast this API does not guarantee it gives back the transfers in order...)

Reply:
ah yup i see